virt-launcher: Use interface for arch specific conversions

jschintag opened the pull request:
Replace long switch case statements for arch specific conversions into separate code files by utilizing an interface.
Fix some code warnings in converter.go

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
Big chunks of arch specific code is mixed with general virt-launcher code
After this PR:
arch specific code is mostly in arch specific files. Exceptions is for trivial paths where only single string values are different and the effort and additional complexity of splitting it would not be worth it.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:
- Some arch specific configuration values are still in the main code
The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->
https://github.com/kubevirt/kubevirt/pull/10490#issuecomment-2024935029
### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

jschintag commented:
/cc @alicefr

alicefr commented on pkg/virt-launcher/virtwrap/converter/archconverter.go:
```diff
@@ -0,0 +1,49 @@
+	isUSBNeeded(c *ConverterContext, vmi *v1.VirtualMachineInstance) bool
+}
+
+func GetArchConverter(arch string) ArchConverter {
```
nit: what about calling this function `NewArchConvertor`?

jschintag replied:
I have renamed the function

alicefr commented on pkg/virt-launcher/virtwrap/converter/converter_test.go:
```diff
@@ -56,7 +56,6 @@ import (
 
 	cmdv1 "kubevirt.io/kubevirt/pkg/handler-launcher-com/cmd/v1"
 	kubevirtpointer "kubevirt.io/kubevirt/pkg/pointer"
-	virtpointer "kubevirt.io/kubevirt/pkg/pointer"
```
Can you please split this formatting/styling changes in a separate commit?

jschintag replied:
I have split the formatting changes into a separate commit.

alicefr commented:
@jschintag the code overall looks good, thanks!  Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?

@andreabolognani @zhlhahaha can you please also take a look to this?

alicefr commented on pkg/virt-launcher/virtwrap/converter/archconverter.go:
```diff
@@ -0,0 +1,49 @@
+		return archConverterARM64{}
+	case isPPC64(arch):
+		return archConverterPPC64{}
+	default:
```
Please add the validation of the arch string.

jschintag replied:
I would rather not do that for 2 reasons:
1. Adding the error makes the parts of the code where we call the functions more complicated, as i can't just slot a call to ArchConverter in as a replacement. Additionally especially `isUSBNeeded` does not throw an error itself. Though this is a minor reason
2. My main reason for not throwing an error is that we do not do that anywhere else in the code. In all cases the default behaviour is always to assume that the arch is x86_64. I don't see a reason to change it here, as it was also here the case

iholder101 replied:
I also think that validation would be valuable here.
@jschintag I would suggest something like:
```go
	switch {
	case isARM64(arch):
		return archConverterARM64{}
	case isPPC64(arch):
		return archConverterPPC64{}
	case isAMD64(arch):
		return archConverterAMD64{}
	default:
		log.Warningf("Trying to create an arch converter from an unknown arch: %s. falling back to AMD64", arch)
		return archConverterAMD64{}
```

This way you will able to do things like `ArchConverter(arch).isUSBNeeded()`, and the default would indeed be AMD64. But, on the other hand, an unknown arch being passed would still be somewhat visible.

My main concern is that without this a bug would probably be very hard to solve.

WDYT?

zhlhahaha commented on pkg/virt-launcher/virtwrap/converter/archconverter.go:
```diff
@@ -0,0 +1,49 @@
+	AddGraphicsDevice(vmi *v1.VirtualMachineInstance, domain *api.Domain, c *ConverterContext)
+	SCSIController(c *ConverterContext, driver *api.ControllerDriver) api.Controller
+	isUSBNeeded(c *ConverterContext, vmi *v1.VirtualMachineInstance) bool
+}
```
Just curious. Why do some functions start with capital letter but one does not?

jschintag replied:
Because i copied the name of the parent `isUSBNeeded` function and forgot to capitalize it. Thanks for catching that.

A different question would be should the interface be exported from the package, since it is not used outside of it?

zhlhahaha replied:
> A different question would be should the interface be exported from the package, since it is not used outside of it?

From my point of view, they are supposed to be private function or interface.

zhlhahaha commented:
Hi @alicefr 
This PR looks good to me. Thanks @jschintag !

fossedihelm commented:
Just an idea: How about encapsulating the `ArchConverter` in a new struct (with dedicated creator) that will replace the
`Architecture` in the converterContext.

```
type ConverterContext struct {
	Architecture                    Architecture
	...
}

type Architecture struct {
	arch string
	ArchConverter
}
```
 In this way, we have to initialize it only once:
 ```
 	c := &converter.ConverterContext{
		Architecture:          converter.NewArchitecture(runtime.GOARCH),
		...
	}
 ```
 
 Maybe in that case we can also add a receiver to the `is<ARC>` functions:
 ```
 func (a *Architecture) IsARM64() bool {
	return a.arch == "arm64"
}

func (a *Architecture) IsPPC64() bool {
	return a.arch == "ppc64le"
}

func (a *Architecture) IsAMD64() bool {
	return a.arch == "amd64"
}

func (a *Architecture) GetArch() string {
	return a.arch
}
 ```
 WDYT?

fossedihelm commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -1209,33 +1209,7 @@ func isUSBNeeded(c *ConverterContext, vmi *v1.VirtualMachineInstance) bool {
 	//In ppc64le usb devices like mouse / keyboard are set by default,
 	//so we can't disable the controller otherwise we run into the following error:
 	//"unsupported configuration: USB is disabled for this domain, but USB devices are present in the domain XML"
```
I think this comment should be moved under the specific method of `IsUSBNeeded` for ppc64le struct

jschintag replied:
Makes sense, i moved it.

jschintag commented:
@fossedihelm I did think about moving the archConverter under the ConverterContext, but decided against it because we do not have a constructor function for ConverterContext, so we would need to change this in every part where ConverterContext is created.
That does not mean we can't do it, it would just be a lot more effort for little benefit, since creating an archConverter is a trivial operation. We would essentially change a code in according to my quick search 40 places accross 6 files just to have the call to archconverter look a bit more cleaner.

jschintag commented:
> @jschintag the code overall looks good, thanks! Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?

@alicefr I have made the split into separate commits as suggested. Regarding the unit tests, i'm not sure what i would need to add here, since all of the functionality of arch converter is already tested explicitly by the unit tests. The only thing that would only be tested by coincidence right now would be `NewArchConverter`. I can add a test for that or did you have more in mind?

alicefr commented:
> > @jschintag the code overall looks good, thanks! Please, split the formatting changes in a separate commit. Would you mind also to add some unit tests for the arch converter?
> 
> @alicefr I have made the split into separate commits as suggested. Regarding the unit tests, i'm not sure what i would need to add here, since all of the functionality of arch converter is already tested explicitly by the unit tests. The only thing that would only be tested by coincidence right now would be `NewArchConverter`. I can add a test for that or did you have more in mind?

You should check that the defaults for each architecture and methods you have added to the interface are the ones you expect. You can overwrite the GOARCH env variable to test the different architectures.

fossedihelm commented:
> @fossedihelm I did think about moving the archConverter under the ConverterContext, but decided against it because we do not have a constructor function for ConverterContext, so we would need to change this in every part where ConverterContext is created. That does not mean we can't do it, it would just be a lot more effort for little benefit, since creating an archConverter is a trivial operation. We would essentially change a code in according to my quick search 40 places accross 6 files just to have the call to archconverter look a bit more cleaner.

Yes, there are several places where we initialize the convertercontex, but most of the time they are part of the unit tests.
And yes, the most important benefit will be readability, but not only that: we can also reduce code duplication playing with it in `Defaulter` struct (which defines `Is<ARCH>` functions again).
We can keep it for a follow up PR, or doing it now. Up to you.
Thank you!

jschintag commented:
Yesterday i have tested adding archConverter to ConverterContext. The main problem is that we would need to ensure that it is always non-empty, as otherwise calling any function that depends on it results in panics.
So even though all calls outside of tests are few and always set Architecture, that is not the  case inside tests, and often times ConverterContext is created ad-hoc.
I think the best way to guarantee that archConverter would not be nil would be to outsource creating it to a function, but i think it would be better to leave this to a follow up PR, since it should be looked at more carefully.

Other than that, i pushed the following changes:
- Added unit-test for newArchConverter function
- made all archConverter calls internal to converter package
- Removed SCSIController function, as it is not needed (yet)
- Simplified Power implementation for graphics device as suggested by https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1601217517

Please re-review

andreabolognani commented:
/lgtm

alicefr commented:
There is a typo in the archconverter_test file name

alicefr commented on pkg/virt-launcher/virtwrap/converter/archvoncerter_test.go:
```diff
@@ -0,0 +1,40 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("archConverter", func() {
```
I would expect here some tests for `addGraphicsDevice` and `isUSBNeeded` too

jschintag replied:
Those are already covered for each arch in `converter_test.go` by comparing them to pre-defined expected xml definitions. As such i don't think it makes much sense to duplicate this code again.
After all the code inside those functions did not change, we just now call it a bit differently. As such it is covered by existing tests.

iholder101 replied:
A bit similar to https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1611613115, the problem as I see it is that code can change. You are right that today `converter_test.go` might be testing it since this PR only brings refactoring and not an actual change of logic, however who knows how the code will look like tomorrow. For example, the arch converter interface can be used in other areas in the future, or the converter logic might change.

The whole concept behind unit tests is that every unit tests itself in a cohesive way. Converter has different tasks, as it converts the vmi into a libvirt XML. While it's true that it uses the arch converter, it uses it as a block box and assumes it's already working. It is not its job to test it and validate it.

One last point is that the most valuable merit of tests in general is NOT to ensure the code is working as expected (this can be validated manually as well, sometimes more easily) but to allow easy refactoring in the future and hence ease future development. Building upon an implicit assumption that the code is being tested by a different unit in an implicit way goes against this principle IMO.

jschintag replied:
I see your point, i think the difference here is what would you think of as a unit? For me in golang that has always been a package, as that is the smallest piece in golang that you can import. As such what a package should be testing is that the exported functionality is stable.
In this case that would be `converter`, as it is the exposed type. On the other hand `archConverter` exists only inside the package, it's only use is to be used by converter as a convenient tool to outsource arch-specific code. So my argument is that if `converter_test.go` does not cover it anymore, it is also no longer needed in the `converter` and should just be removed.

iholder101 replied:
That's a fair argument.
I tend to see a unit as the smaller form of an "entity" that provides some kind of a service. The reason it is valuable to look at is this way is that even inside the package, this interface can end up being used for something completely different. If we don't see the interface as a unit, then it remains untested in this case.

But in any case, I don't think it should be blocking for this PR.

jschintag commented on pkg/virt-launcher/virtwrap/converter/BUILD.bazel:
```diff
@@ -45,6 +50,7 @@ go_library(
 go_test(
     name = "go_default_test",
     srcs = [
+        "archvoncerter_test.go",
```
fixed

jschintag commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -1206,36 +1206,7 @@ func initializeQEMUCmdAndQEMUArg(domain *api.Domain) {
-	}
-
-	return false
+	return newArchConverter(c.Architecture).isUSBNeeded(c, vmi)
```
I have removed the function and called archConverter directly.

fossedihelm commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -88,6 +88,11 @@ const (
 	MemoryHotplugBlockAlignmentBytes = 2097152
 )
 
+const (
+	graphicsDeviceDefaultHeads uint = 1
+	graphicsDeviceDefaultVRAM  uint = 16384
+)
+
```
just an idea:
Maybe we can move this inside `archconverter.go` file

jschintag replied:
I have moved it. It can be argued that it makes sense in both files, but in the end the archconverter is using it.

fossedihelm reviewed (commented):
Thank you for your patience.
We are almost there, just small minor comments.

jschintag commented on pkg/virt-launcher/virtwrap/converter/archconverter.go:
```diff
@@ -0,0 +1,39 @@
+
+type archConverter interface {
+	addGraphicsDevice(vmi *v1.VirtualMachineInstance, domain *api.Domain, c *ConverterContext)
+	isUSBNeeded(c *ConverterContext, vmi *v1.VirtualMachineInstance) bool
```
I have dropped it from `isUSBNeeded`. I have left it in `addGraphicsDevice`, since even though only amd64 needs a single boolean from it, giving the function a pointer is just as cheap as copying the boolean.

fossedihelm replied:
Fair enough, thanks

fossedihelm reviewed (approved):
Thank you for this!
Nothing else from my side.
Good job

jschintag commented:
/assign @iholder101

alicefr commented:
/approve
/hold
Put on hold if @iholder101 wants to take a look, otherwise, feel free to unhold
@jschintag thanks for the work!

iholder101 commented on pkg/virt-launcher/virtwrap/converter/arm64.go:
Can you add unit tests for the three converters?

jschintag replied:
See https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1603027526

iholder101 replied:
Thanks for the pointer :+1: Let's discuss it there

iholder101 commented on pkg/virt-launcher/virtwrap/converter/arm64.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+type archConverterARM64 struct{}
+
```
Can you please add a `var _ = archConverter(&archConverterARM64{})`?
This will ensure in compilation time that the struct implements the interface and fail otherwise.
Please do so for all 3 converters.

jschintag replied:
The `newArchConverter` function already explicitly returns `archConverter`, so this is already covered.
For example if you comment out isUSBNeeded:
```
pkg/virt-launcher/virtwrap/converter/archconverter.go:40:10: cannot use archConverterARM64{} (value of type archConverterARM64) as archConverter value in return statement: archConverterARM64 does not implement archConverter (missing method isUSBNeeded)
```

iholder101 replied:
I see your point.

My problem with it is that it's implicit and hard to understand for someone not familiar enough with the code. For example, when looking at this file, how can I know that it aims to implement an interface? In this case it's even more important, as this struct's whole purpose (at least currently) is to implement this interface.

I think that adding it here will be valuable as it will make the code more cohesive, clearer and self-contained. But I'm willing to be convinced otherwise.

Is there a reason you think it's more valuable to not have this validation here? Are there any downsides to it?

jschintag replied:
No that makes sense.
The only real downside i see would be people wondering why the line is there. But i have just added a comment to it for that.

iholder101 reviewed (commented):
Thank you @jschintag! Looks great!
See comments below

jschintag commented:
@iholder101 Thank you for your review.
Github does not let me quote https://github.com/kubevirt/kubevirt/pull/11907#discussion_r1611334858, but i liked the idea and have implemented it. Though i used the logger from client-go, as it is already widely used in the project.

Would you mind reviewing the changes again?

iholder101 commented on pkg/virt-launcher/virtwrap/converter/archconverter.go:
```diff
@@ -0,0 +1,49 @@
+	case isAMD64(arch):
+		return archConverterAMD64{}
+	default:
+		log.Log.Warning(("Trying to create an arch converter from an unknown arch: " + arch + ". falling back to AMD64"))
```
typo
```suggestion
		log.Log.Warning("Trying to create an arch converter from an unknown arch: " + arch + ". falling back to AMD64")
```

jschintag replied:
fixed, most likely that was vscode trying to be helpful

iholder101 commented on pkg/virt-launcher/virtwrap/converter/archconverter_test.go:
```diff
@@ -0,0 +1,36 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("archConverter", func() {
```
nit
```suggestion
var _ = Describe("Arch Converter", func() {
```

jschintag replied:
I have changed it

iholder101 commented:
Thanks a lot @jschintag! Great stuff!
/lgtm
/hold cancel
